src/language_server.ts

```diff
@@ -619,33 +627,46 @@ export class LanguageServer {
 			log("restarting Language Server");
 		}
 		const enable: boolean = workspace.getConfiguration('inmanta').ls.enabled;
-		await this.stopServerAndClient();
-		if (enable) {
-			await this.startServerAndClient();
-			window.showInformationMessage(`The Language server has been enabled for folder ${this.rootFolder.name}`);
-		}
-
+		await this.mutex.runExclusive(async () => {
+			await this.stopServerAndClient(false);
+			if (enable) {
+				await this.startServerAndClient();
+				window.showInformationMessage(`The Language server has been enabled for folder ${this.rootFolder.name}`);
+			}
+		});
 	}
 
 	/**
 	 * Stops the language server and its client.
 	 */
-	async stopServerAndClient() {
+	async stopServerAndClient(acquire_lock: boolean = true) {
 		log("Stopping server and client...");
-		await this.mutex.runExclusive(async () => {
-			if (this.client) {
-				if(this.client.needsStop()){
-					await this.client.stop();
-				}
-				this.client = undefined;
+		if(acquire_lock){
+			await this.mutex.runExclusive(async () => {
+				await this._stopServerAndClient();
+			});
+		} else {
+			await this._stopServerAndClient();
+		}
+	}
+
+	/* 
+	 * Stops the language server and its client.
+	 * This function should always run under the `mutex` lock.
```

Comment:
This says "always", yet `stopServerAndClient` has a boolean to control it.

Reply:
This docstring belongs to the `_stopServerAndClient()` method and not to the `stopServerAndClient` method

Reply:
Yes, I meant to say that `stopServerAndClient`'s boolean controls whether or not it calls `_stopServerAndClient` under the mutex lock. So it violates the docstring of this method. Either the docstring is too strict, or the boolean should not be there in the caller method.

Reply:
Or should the boolean be false only if it is already running under the lock?

Reply:
yes